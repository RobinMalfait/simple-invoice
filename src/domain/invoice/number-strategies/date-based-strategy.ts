import { format } from 'date-fns'

export const Base = {
  yyyyMMdd: (date: Date) => format(date, 'yyyyMMdd'),
  yyyyQ: (date: Date) => format(date, 'yyyyQ'),
  yyQ: (date: Date) => format(date, 'yyQ'),
  yyQQ: (date: Date) => format(date, 'yyQQ'),
  yy: (date: Date) => format(date, 'yy'),
}

export class DateBasedStrategy {
  private countsByDate = new Map<string, number>()

  public constructor(
    private base = Base.yyyyMMdd,
    private significantDigits = 3,
  ) {}

  public next = (issueDate: Date): string => {
    let base = this.base(issueDate)

    if (this.countsByDate.has(base)) {
      this.countsByDate.set(base, this.countsByDate.get(base)! + 1)
    } else {
      this.countsByDate.set(base, 1)
    }

    let postfix = `${this.countsByDate.get(base)}`.padStart(this.significantDigits, '0')
    return `${base}${postfix}`
  }
}
