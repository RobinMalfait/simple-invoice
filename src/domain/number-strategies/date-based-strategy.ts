import { format } from 'date-fns'

export let Base = {
  yyyyMMdd: (date: Date) => {
    return format(date, 'yyyyMMdd')
  },
  yyyyQ: (date: Date) => {
    return format(date, 'yyyyQ')
  },
  yyQ: (date: Date) => {
    return format(date, 'yyQ')
  },
  yyQQ: (date: Date) => {
    return format(date, 'yyQQ')
  },
  yy: (date: Date) => {
    return format(date, 'yy')
  },
}

export class DateBasedStrategy {
  private countsByDate = new Map<string, number>()

  public constructor(
    private base = Base.yyyyMMdd,
    private significantDigits = 3,
  ) {}

  public next = (date: Date): string => {
    let base = this.base(date)

    if (this.countsByDate.has(base)) {
      this.countsByDate.set(base, this.countsByDate.get(base)! + 1)
    } else {
      this.countsByDate.set(base, 1)
    }

    let postfix = `${this.countsByDate.get(base)}`.padStart(this.significantDigits, '0')
    return `${base}${postfix}`
  }
}
