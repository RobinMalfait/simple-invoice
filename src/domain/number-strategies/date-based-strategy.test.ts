import { parseISO } from 'date-fns'
import { Base, DateBasedStrategy } from './date-based-strategy'

it('should calculate the next invoice number for the same date', () => {
  let strategy = new DateBasedStrategy()
  let now = parseISO('2023-02-01')

  expect(strategy.next(now)).toEqual('20230201001')
  expect(strategy.next(now)).toEqual('20230201002')
})

it.each(
  Object.entries(Base).map(([name, fn]) => {
    return [name, fn]
  }),
)(
  'should calculate the next invoice number for the same date with a custom base "%s"',
  (_, base) => {
    let strategy = new DateBasedStrategy(base)
    let now = parseISO('2023-02-01')

    expect(strategy.next(now)).toMatchSnapshot()
    expect(strategy.next(now)).toMatchSnapshot()
  },
)

it.each(
  Object.entries(Base).map(([name, fn]) => {
    return [name, fn]
  }),
)(
  'should calculate the next invoice number for the same date with a custom base "%s", and with a different amount of significantDigits',
  (_, base) => {
    let strategy = new DateBasedStrategy(base, 2)
    let now = parseISO('2023-02-01')

    expect(strategy.next(now)).toMatchSnapshot()
    expect(strategy.next(now)).toMatchSnapshot()
  },
)
