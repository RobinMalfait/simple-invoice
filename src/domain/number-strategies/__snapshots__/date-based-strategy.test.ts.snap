// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should calculate the next invoice number for the same date with a custom base "yy" 1`] = `"23001"`;

exports[`should calculate the next invoice number for the same date with a custom base "yy" 2`] = `"23002"`;

exports[`should calculate the next invoice number for the same date with a custom base "yy", and with a different amount of significantDigits 1`] = `"2301"`;

exports[`should calculate the next invoice number for the same date with a custom base "yy", and with a different amount of significantDigits 2`] = `"2302"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyQ" 1`] = `"231001"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyQ" 2`] = `"231002"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyQ", and with a different amount of significantDigits 1`] = `"23101"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyQ", and with a different amount of significantDigits 2`] = `"23102"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyQQ" 1`] = `"2301001"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyQQ" 2`] = `"2301002"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyQQ", and with a different amount of significantDigits 1`] = `"230101"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyQQ", and with a different amount of significantDigits 2`] = `"230102"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyyyMMdd" 1`] = `"20230201001"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyyyMMdd" 2`] = `"20230201002"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyyyMMdd", and with a different amount of significantDigits 1`] = `"2023020101"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyyyMMdd", and with a different amount of significantDigits 2`] = `"2023020102"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyyyQ" 1`] = `"20231001"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyyyQ" 2`] = `"20231002"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyyyQ", and with a different amount of significantDigits 1`] = `"2023101"`;

exports[`should calculate the next invoice number for the same date with a custom base "yyyyQ", and with a different amount of significantDigits 2`] = `"2023102"`;
